### DISCUSSION TITLE
Proper way to implement paging

### DISCUSSION THREAD

**Discussion by jibla** [neet/masto.js#993] [Q&A] [answered]
Pagination parameters (which is extended then) looks like folloowing:
```typescript
interface DefaultPaginationParams {
    /** Return results older than this ID. */
    readonly maxId?: string | null;
    /** Return results newer than this ID. */
    readonly sinceId?: string | null;
    /** Get a list of items with ID greater than this value excluding this ID */
    readonly minId?: string | null;
    /** Maximum number of results to return per page. Defaults to 40. NOTE: Pagination is done with the Link header from the response. */
    readonly limit?: number | null;
}
```

Also, based on mastodon api docs, we know that parameters like `maxId`, `sinceId` and `minId` are used for controlling pagination. However, in the comment above the `limit`, it is said to take in mind pagination is done using Link header.

My question is, what is the proper way to implement paging with masto.js? If I use `minId` and change it appropriately when calling for example .list() method, is not it enough? Does the library support those mentioned Link headers? Or is there any other way I have to follow?

**Comment by neet** [comment, marked as answer]
Hello @jibla. As you pointed out, Mastodon returns information related to pagination including the next `minId` in the HTTP `Link` header, and Masto.js is currently not offering a API for directly referring response headers.

However, we instead offer a higher level abstraction. All methods named `list()` actually returns an object that we call [`Paginator`](https://neet.github.io/masto.js/interfaces/mastodon.Paginator.html) that can not only behave like a `Promise`, but also as a [`AsyncIterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols).

Here's a basic example that favourite all posts on your home feed by using [`for-await-of` syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of):

```ts
for await (const statuses of masto.v1.timelines.home.list()) {
  for (const status of statuses) {
    await masto.v1.statuses.$select(status.id).favourite();
  }
}
```

or if you want to fetch individual pages, you can use `next()`

```ts
const timeline = masto.v1.timelines.home.list();

const page1 = await timeline.next() // { done: false, value: [post, post, post...] } 
const page2 = await timeline.next();
const page3 = await timeline.next();
```

at the same time, you can fetch the first page by `await`

```ts
const timeline = await masto.v1.timelines.home.list();
//               ^^^^^
//               It now behaves like a Promise because we used `await`.
//               This is a synonym for `(await timelines.home.list().next()).value`
```

Therefore I believe we don't need to expose any HTTP information because you can encapsulate concrete implementation by using the Paginator API.

TSDoc comment for the limit property mentions about Link header but this is because we copy-and-pasted all descriptions from the Mastodon docs, which should be improved in the future.

I hope this helps

**Reply by jibla** [reply]
@neet Thank you!
